File: php/54_Spiral_Matrix.php
<?php

class Solution {

    /**
     * @param Integer[][] $matrix
     * @return Integer[]
     */
    function spiralOrder($matrix) {
        $arResult = [];
        if (empty($matrix)) {
            return $arResult;
        }

        $rowStart = 0;
        $rowEnd = count($matrix) -  1;
        $columnStart = 0;
        $columnEnd = count($matrix[0]) - 1;

        while ($rowStart <= $rowEnd && $columnStart <= $columnEnd) {
            for ($i = $columnStart; $i <= $columnEnd; $i++) {
                $arResult[] = $matrix[$rowStart][$i];
            }
            $rowStart++;
            for ($i = $rowStart; $i <= $rowEnd; $i++) {
                $arResult[] = $matrix[$i][$columnEnd];
            }
            $columnEnd--;
            if ($rowStart <= $rowEnd) {
                for ($i = $columnEnd; $i >= $columnStart; $i--) {
                    $arResult[] = $matrix[$rowEnd][$i];
                }
            }
            $rowEnd--;
            if ($columnStart <= $columnEnd) {
                for ($i = $rowEnd; $i >= $rowStart; $i--) {
                    $arResult[] = $matrix[$i][$columnStart];
                }
            }
            $columnStart++;
        }

        return $arResult;
    }
}

File: go/54_Spiral_Matrix.go
func spiralOrder(matrix [][]int) []int {
    var result []int
    if len(matrix) == 0 {
        return result
    }

    var rowStart = 0
    var rowEnd = len(matrix) -  1
    var columnStart = 0
    var columnEnd = len(matrix[0]) - 1

    for rowStart <= rowEnd && columnStart <= columnEnd {
        for i := columnStart; i <= columnEnd; i++ {
            result = append(result, matrix[rowStart][i]);
        }
        rowStart++;
        for i := rowStart; i <= rowEnd; i++ {
            result = append(result, matrix[i][columnEnd]);
        }
        columnEnd--;
        if rowStart <= rowEnd {
            for i := columnEnd; i >= columnStart; i-- {
                result = append(result, matrix[rowEnd][i]);
            }
        }
        rowEnd--;
        if columnStart <= columnEnd {
            for i := rowEnd; i >= rowStart; i-- {
                result = append(result, matrix[i][columnStart]);
            }
        }
        columnStart++;
    }

    return result;
}